File: arrayAlgorithms.cpp
#include <bits/stdc++.h>
using namespace std;

typedef struct
{
    int id;
    int solved;
    int penalty;
} team;

bool icpcCmp(team a, team b)
{
    if (a.solved != b.solved)
        return a.solved > b.solved;
    else if (a.penalty != b.penalty)

        return a.penalty < b.penalty;
    else
        return a.id < b.id;
}

int main()
{
    int arr[] = {10, 7, 2, 15, 4};
    vector<int> v(arr, arr + 5); // alternative constructor

    // sort descending with vector
    sort(v.rbegin(), v.rend()); // reverse iterator
    for (auto &val : v)
        printf("%d ", val); // access the values
    printf("\n");
    printf("==================\n");

    // sort descending with integer array
    sort(arr, arr + 5);    // ascending
    reverse(arr, arr + 5); // then reverse
    for (int i = 0; i < 5; ++i)
        printf("%d ", arr[i]);
    printf("\n");
    printf("==================\n");

    shuffle(v.begin(), v.end(), default_random_engine()); // shuffle the content
    for (auto &val : v)
        printf("%d ", val);
    printf("\n");
    printf("==================\n");
    partial_sort(v.begin(), v.begin() + 2, v.end()); // partial_sort demo
    for (auto &val : v)
        printf("%d ", val);
    printf("\n");
    printf("==================\n");

    // sort ascending
    sort(arr, arr + 5);         // arr is sorted now
    for (int i = 0; i < 5; ++i) // 2, 4, 7, 10, 15
        printf("%d ", arr[i]);
    printf("\n");
    sort(v.begin(), v.end()); // sort a vector
    for (auto &val : v)       // 2, 4, 7, 10, 15 too
        printf("%d ", val);
    printf("\n");
    printf("==================\n");

    // multi-field sorting example, suppose we have 4 ICPC teams
    team nus[4] = {{1, 1, 10},
                   {2, 3, 60},
                   {3, 1, 20},
                   {4, 3, 60}};

File: arrayAlgorithms.java
import java.util.*;

class team implements Comparable<team> {
    private int id, solved, penalty;

    public team(int i, int solved, int penalty) {
        this.id = id;
        this.solved = solved;
        this.penalty = penalty;
    }

    public int compareTo(team o) {
        if (solved != o.solved)
            return o.solved - solved;
        else if (penalty != o.penalty)

            return penalty - o.penalty;
        else
            return id - o.id;

    }

    public String toString() {
        return "id: " + id + ", solved: " + solved + ", penalty: " + penalty;
    }
}

class arrayAlgorithms {
    public static void main(String[] args) {
        ArrayList<Integer> v = new ArrayList<>();

        v.add(10);
        v.add(7);
        v.add(2);
        v.add(15);
        v.add(4);

        // sort descending with vector
        Collections.sort(v);
        // if we want to modify comparison function, use the overloaded method:
        // Collections.sort(List list, Comparator c);
        Collections.reverse(v);

        System.out.println(v);
        System.out.printf("==================\n");

        // shuffle the content again
        Collections.shuffle(v);
        System.out.println(v);
        System.out.printf("==================\n");

        // sort ascending
        Collections.sort(v);
        System.out.println(v);
        System.out.printf("==================\n");

        ArrayList<team> nus = new ArrayList<>();
        nus.add(new team(1, 1, 10));
        nus.add(new team(2, 3, 60));
        nus.add(new team(3, 1, 20));
        nus.add(new team(4, 3, 60));

        // without sorting, they will be ranked like this:
        for (int i = 0; i < 4; ++i)
            System.out.println(nus.get(i));

        Collections.sort(nus); // sort using a comparison function
        System.out.printf("==================\n");
        // after sorting using ICPC rule, they will be ranked like this:
        for (int i = 0; i < 4; ++i)
            System.out.println(nus.get(i));
        System.out.printf("==================\n");

        int pos = Collections.binarySearch(v, 7);
        System.out.println("Trying to search for 7 in v, found at index = " + pos);

        pos = Collections.binarySearch(v, 77);
        System.out.println("Trying to search for 77 in v, found at index = " + pos);
    }

}